### ISSUE TITLE
2.5.1 MatlabLexer regression

### ISSUE THREAD

**Issue by florczakraf** [pygments/pygments#1327]
https://github.com/pygments/pygments/pull/1271 introduced a regression to `MatlabLexer`.

### Reproduction script:
```
import pygments
from pygments.lexers.matlab import MatlabLexer

print(list(pygments.lex('foo bar', MatlabLexer())))
print(list(pygments.lex('foo', MatlabLexer())))
```

### Before (2.5.1 with 612fb2b reverted):
```
[(Token.Name, 'foo'), (Token.Text, ' '), (Token.Name, 'bar'), (Token.Text, '\n')]
[(Token.Name, 'foo'), (Token.Text, '\n')]
```

### After:
There are invalid characters (` `, `\n`) in `Name`. ~~Moreover, `bar` is of a wrong type now.~~
```
[(Token.Name, 'foo '), (Token.Literal.String, 'bar'), (Token.Punctuation, '\n')]
[(Token.Name, 'foo\n'), (Token.Punctuation, '')]
```

@birkenfeld @anntzer could you please have another look at those changes?

**Comment by birkenfeld** [comment]
I believe the change in `bar` was intended by @anntzer, since this is "command syntax" with arguments.

The `Name` should be fixed to not include the newline, of course.

**Comment by anntzer** [comment]
"bar" is effectively a string from MATLAB's POV (`foo bar` is equivalent to `foo('bar')`) and I would say should be highlighted as such.
I guess the invalid characters can be fixed by making the \s+ in https://github.com/pygments/pygments/commit/612fb2b33d6f48965661e53c6a88ec78b5dd54e3#diff-bc1be5d33c811ece546521ae70944206R104 part of the lookahead and by removing the newline from https://github.com/pygments/pygments/commit/612fb2b33d6f48965661e53c6a88ec78b5dd54e3#diff-bc1be5d33c811ece546521ae70944206R146?

**Comment by birkenfeld** [comment]
SGTM.

**Comment by florczakraf** [comment]
> "bar" is effectively a string from MATLAB's POV (foo bar is equivalent to foo('bar')) and I would say should be highlighted as such.

You are absolutely right here! I'm still a little bit confused with matlab's calling conventions...

**Comment by anntzer** [comment]
Wait now I'm confused, in https://github.com/pygments/pygments/commit/612fb2b33d6f48965661e53c6a88ec78b5dd54e3#diff-bc1be5d33c811ece546521ae70944206R146 the \n is in a negated group so it's not clear to me why the parsing ends up including the \n...

**Comment by birkenfeld** [comment]
True. I guess it's only the first location you mentioned, then.

**Comment by anntzer** [comment]
@florczakraf Can you check whether
```patch
diff --git i/pygments/lexers/matlab.py w/pygments/lexers/matlab.py
index ee85d088..befccd85 100644
--- i/pygments/lexers/matlab.py
+++ w/pygments/lexers/matlab.py
@@ -101,7 +101,7 @@ class MatlabLexer(RegexLexer):
             # is recognized if it is either surrounded by spaces or by no
             # spaces on both sides; only the former case matters for us.  (This
             # allows distinguishing `cd ./foo` from `cd ./ foo`.)
-            (r'(?:^|(?<=;))\s*\w+\s+(?!=|\(|(%s)\s+)' % _operators, Name,
+            (r'(?:^|(?<=;))\s*\w+(?=\s)(?!\s+(=|\(|(%s)\s+))' % _operators, Name,
              'commandargs'),
 
             # operators:
@@ -142,9 +142,10 @@ class MatlabLexer(RegexLexer):
             (r"[^']*'", String, '#pop'),
         ],
         'commandargs': [
+            ("\s+", Text),
             ("'[^']*'", String),
             ("[^';\n]+", String),
-            (";?\n?", Punctuation, '#pop'),
+            (";?(?=\n)?", Punctuation, '#pop'),
         ]
     }
```
works for you?

**Comment by florczakraf** [comment]
I think it doesn't solve a problem of preceding whitespace:

```
>>> list(pygments.lex('    foo    ', MatlabLexer()))
[(Token.Name, '    foo'), (Token.Text, '    \n'), (Token.Punctuation, '')]
```

**Comment by anntzer** [comment]
New try:
```patch
diff --git i/pygments/lexers/matlab.py w/pygments/lexers/matlab.py
index ee85d088..19ed89ae 100644
--- i/pygments/lexers/matlab.py
+++ w/pygments/lexers/matlab.py
@@ -101,7 +101,7 @@ class MatlabLexer(RegexLexer):
             # is recognized if it is either surrounded by spaces or by no
             # spaces on both sides; only the former case matters for us.  (This
             # allows distinguishing `cd ./foo` from `cd ./ foo`.)
-            (r'(?:^|(?<=;))\s*\w+\s+(?!=|\(|(%s)\s+)' % _operators, Name,
+            (r'(?:^|(?<=;))(\s*)(\w+)(?=\s)(?!\s+(=|\(|(%s)\s+))' % _operators, bygroups(Text, Name),
              'commandargs'),
 
             # operators:
@@ -142,9 +142,10 @@ class MatlabLexer(RegexLexer):
             (r"[^']*'", String, '#pop'),
         ],
         'commandargs': [
+            ("\s+", Text),
             ("'[^']*'", String),
             ("[^';\n]+", String),
-            (";?\n?", Punctuation, '#pop'),
+            (";?(?=\n)?", Punctuation, '#pop'),
         ]
     }

```

**Comment by florczakraf** [comment]
LGTM (`Token.Name` are now properly separated) but now we have a side effect of empty trailing `Token.Punctuation`: `... (Token.Text, '\n'), (Token.Punctuation, '')]` which doesn't actually cause any issues in my use case

**Comment by birkenfeld** [comment]
fixed in #1328
